_archive/blech_run_auto_process.py

```diff
@@ -0,0 +1,369 @@
+electrode_num_list.sort()
+
+for electrode_num in electrode_num_list:
+
+	# electrode_output_dir = os.path.join(autosort_output_dir, f'electrode{electrode_num:02}')
+
+	############################################################
+	# Get unit details and load data
+	############################################################
+
+	print()
+	print('======================================')
+	print()
+
+	# Iterate over electrodes and pull out spikes
+	# Get classifier probabilities for each spike and use only
+	# "good" spikes
+
+	# Print out selections
+	print(f'=== Processing Electrode {electrode_num:02} ===')
+
+
+	# Load data from the chosen electrode and solution
+	# We can simply load data from 2-clusters because we're not using
+	# gmm-predictions
+	(
+		spike_waveforms,
+		spike_times,
+		pca_slices,
+		energy,
+		amplitudes,
+		predictions,
+		) = post_utils.load_data_from_disk(electrode_num, 2)
+
+	clf_data_paths = [
+			f'./spike_waveforms/electrode{electrode_num:02}/clf_prob.npy',
+			f'./spike_waveforms/electrode{electrode_num:02}/clf_pred.npy',
+			]
+	clf_prob, clf_pred = [np.load(this_path) for this_path in clf_data_paths]
+
+	# Toss all waveforms and times that are not predicted
+	spike_waveforms = spike_waveforms[clf_pred]
+	spike_times = spike_times[clf_pred]
+	pca_slices = pca_slices[clf_pred]
+	energy = energy[clf_pred]
+	amplitudes = amplitudes[clf_pred]
+	clf_prob = clf_prob[clf_pred]
+	clf_pred = clf_pred[clf_pred]
+
+	# Take everything
+	data = post_utils.prepare_data(
+			np.arange(len(spike_waveforms)),
+			pca_slices,
+			energy,
+			amplitudes,
+			)
+
+	############################################################
+	# Perform clustering
+	############################################################
+
+	# Since the dirichlet process WILL find more clusters for more data,
+	# (whereas we know that more spikes doens't mean more neurons),
+	# we need to keep the "COUNT" constant
+	# One way to do this is to use KMeans centroids as data points
+
+	max_k = 1000
+	n_max_train = 10000
+	if len(data) > max_k:
+		kmean_obj = KMeans(n_clusters = max_k)	
+		if len(data) < n_max_train:
+			kmean_obj.fit(data)
+		else:
+			inds = np.random.choice(len(data), n_max_train, replace = False)
+			kmean_obj.fit(data[inds])
+		train_data = kmean_obj.cluster_centers_
+		kmeans_used = True
+	else:
+		train_data = data
+		kmeans_used = False
+
+	max_components = 10
+	g = BGM(
+			n_components = max_components,
+			covariance_type = 'full',
+			max_iter = 1000,
+			n_init = 10,
+			weight_concentration_prior_type = 'dirichlet_process',
+			# This can be systematically adjusted to match
+			# actual data
+			weight_concentration_prior = 0.1,
+			verbose = 10,
+			)
+
+	#n_max_train = 20000
+	#if len(data) < n_max_train:
+	#	g.fit(data)
+	#else:
+	#	inds = np.random.choice(len(data), n_max_train, replace = False)
+	#	g.fit(data[inds])
+
+	# Train on kmeans centroids but predict on the actual data
+	g.fit(train_data)
```

Comment:
in case of dataset length < max_k, you're fitting and testing on the same data - could this be a problem?

Reply:
That should be fine. I guess the `training` set really isn't is a training set...it's more like a subset. So if we can fit to the full dataset, that's great...otherwise we make the subset (training set).